### PR TITLE
ENH: add parameter `strict` to `assert_allclose`

### PR DESCRIPTION
**Pull request by mdhaber**
This PR adds the parameter `strict` to `np.testing.assert_allclose`. With `strict=True`, the shape and dtype of the two arguments must match for the assertion to pass.

Follows the example of gh-21595
Partially addresses suggestion in https://github.com/numpy/numpy/pull/24667#discussion_r1319991254

### CODE REVIEW

File: numpy/testing/_private/utils.py [hunk 1]
```diff
@@ -1484,13 +1490,47 @@ def assert_allclose(actual, desired, rtol=1e-7, atol=0, equal_nan=True,
...
+
+    Use `strict` to raise an ``AssertionError`` when comparing an array
+    with one or more dimensions against a scalar.
+
+    >>> np.testing.assert_allclose(np.sin(x), 0, atol=1e-15, strict=True)
+    Traceback (most recent call last):
+        ...
+    AssertionError:
+    Not equal to tolerance rtol=1e-07, atol=1e-15
+
```
Comment (mdhaber) [inline numpy/testing/_private/utils.py:1516]:
```suggestion
<BLANKLINE>
```

File: numpy/testing/_private/utils.py [hunk 2]
```diff
@@ -1484,13 +1490,47 @@ def assert_allclose(actual, desired, rtol=1e-7, atol=0, equal_nan=True,
...
+    The `strict` parameter also ensures that the array data types match:
+
+    >>> y = np.zeros(3, dtype=np.float32)
+    >>> np.testing.assert_allclose(np.sin(x), y, atol=1e-15, strict=True)
+    Traceback (most recent call last):
+        ...
+    AssertionError:
+    Not equal to tolerance rtol=1e-07, atol=1e-15
+
```
Comment (mdhaber) [inline numpy/testing/_private/utils.py:1529]:
```suggestion
<BLANKLINE>
```

**Review by mdhaber — COMMENTED** [review]
To fix the CircleCI failure.

**Comment by mdhaber** [comment]
IIUC, I should add a file `doc/release/upcoming_changes/24680.new_feature.rst` with contents like
```
``strict`` option for `testing.assert_allclose`
-----------------------------------------------
The ``strict`` option is now available for `testing.assert_allclose`.
Setting ``strict=True`` will disable the broadcasting behaviour for scalars and
ensure that input arrays have the same data type.
```

**Comment by seberg** [comment]
That seems about right, yes.

File: numpy/testing/_private/utils.py [hunk 3]
```diff
@@ -1484,13 +1490,47 @@ def assert_allclose(actual, desired, rtol=1e-7, atol=0, equal_nan=True,
...
+    This behaviour can be disabled with the `strict` parameter.
 
     Examples
     --------
     >>> x = [1e-5, 1e-3, 1e-1]
     >>> y = np.arccos(np.cos(x))
     >>> np.testing.assert_allclose(x, y, rtol=1e-5, atol=0)
 
+    As mentioned in the Notes section, `assert_allclose` has special
+    handling for scalars. Here, the test checks that the value of `np.sin`
```
Comment (mdhaber) [inline numpy/testing/_private/utils.py:1502]: To make the link render as intended:
```suggestion
    handling for scalars. Here, the test checks that the value of `numpy.sin`
```

File: numpy/testing/_private/utils.py [hunk 4]
```diff
@@ -1484,13 +1490,47 @@ def assert_allclose(actual, desired, rtol=1e-7, atol=0, equal_nan=True,
...
+    is nearly zero at integer multiples of pi.
```
Comment (mdhaber) [inline numpy/testing/_private/utils.py:1503]: If symbols are allowed as they are in SciPy, I'd prefer:
```suggestion
    is nearly zero at integer multiples of π.
```

**Review by mdhaber — COMMENTED** [review]
Two minor tweaks for the reviewer to consider. Feel free to commit.
When this looks good, I'll submit a follow-up for `assert_equal`, if that would be helpful?
These will help address https://github.com/numpy/numpy/pull/24667#discussion_r1319593999.

**Comment by mattip** [comment]
@mdhaber Please go ahead and commit your self-review changes so we can get this in.

**Comment by mdhaber** [comment]
Thanks @mattip. Shall I do the same for `assert_equal`?

**Comment by mattip** [comment]
Thanks @mdhaber. Yes, I think `assert_equal` could also get a `strict` kwarg. At the same time, the docstring of `assert_equal` could use some touchup: it is missing the Notes and See Also that the others have.

**Comment by mdhaber** [comment]
@mattip I opened gh-24770 to add the `strict` kwarg to `assert_equal` and touch up the docs as you suggested. gh-24775  adds `strict` to `assert_array_less`.

**Pull request merged**